Update region hint enum values to make them sequential

avilcheslopez opened the pull request:
Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

- Fixes #1547 

Updated geopm_region_hint_e to make the enum values sequential, updated all code using mappings involving this enum to use array indexing, instead (where appropriate).

avilcheslopez committed:
Update region hint enum values to make them sequential

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez commented on service/src/Helper.cpp:
```diff
@@ -210,17 +210,27 @@ namespace geopm
         }
     }
 
+    unsigned int hint_to_index(uint64_t hint)
```
This is perhaps not the best file for this function, but placed it here for now. Wanted to get a draft PR out there for now. Will revisit this in the meantime.

avilcheslopez replied:
A bit of context: I decided to put it there for now to follow the current pattern, as I noticed another hint enum helper function was placed there as well (``check_hint()``).

However, thinking about it a bit more, I would propose these functions (``check_hint()``, ``hint_to_index()``) be moved to ``geopm_hint.h`` for cohesion. The functions in ``Helper.hpp`` are more general-purpose, while the hint functions are directly tied to the hint enum. Also, keeping them together with the enum will make them easier to maintain, and you can reduce header dependencies as code that needs to use the enum will only need to include one header (as opposed to two).

cmcantalupo replied:
sounds good!

avilcheslopez committed:
Misc. updates for changes to region hint enum

- Added unit tests for hint helper functions
- Updated Fortran code accordingly (as it also defines the region
  hint enum)

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

cmcantalupo commented on service/src/geopm_hint.h:
```diff
@@ -17,14 +17,15 @@ extern "C" {
-    GEOPM_REGION_HINT_PARALLEL =  1ULL << 38, /* Region is threaded */
-    GEOPM_REGION_HINT_IGNORE =    1ULL << 39, /* Do not add region time to epoch */
-    GEOPM_REGION_HINT_INACTIVE =  1ULL << 40, /* Used to marke CPUs that are not running application */
+    GEOPM_REGION_HINT_COMPUTE =   2ULL << 32, /* Region dominated by compute */
```
I think hint can actually range from 0-N (i.e. I don't think we need the 32 bit shift)? There should no longer be a "hint" embedded in the upper 32 bits of the "region_id" but maybe I have forgotten something.

avilcheslopez replied:
I opted to still do the shifting because it is still being embedded in region_id. See: ``Profile.cpp``. If the plan is to move away from that, I can look into it.

avilcheslopez committed:
Isolate bit-shifting logic of hint enum to region_id

Make hint enum truly sequential (no longer << 32), isolate the
bit-shifting logic to only the code that deals with region_id (where
hint enums are embedded).

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Move hint helper function to geopm_hint.h

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Add message to static_asserts in record.cpp

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Add license header to GEOPMHint.cpp, GEOPMHintTest.cpp

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

cmcantalupo commented on service/src/GEOPMHint.cpp:
```diff
@@ -0,0 +1,31 @@
+#include "geopm/Helper.hpp"
+#include "geopm/Exception.hpp"
+
+extern "C" {
```
```suggestion
namespace geopm
{
```

avilcheslopez replied:
Considering the header is a C header, were you thinking of converting it to a C++ header, or is it that you just want to put ``check_hint()`` in the geopm namespace?

cmcantalupo commented on service/src/geopm_hint.h:
```diff
@@ -15,28 +15,31 @@ extern "C" {
+                                GEOPM_REGION_HINT_INACTIVE,
 };
 
+/// @brief Verify a region's hint value is legal for use.
+void check_hint(uint64_t hint);
+
 #ifdef __cplusplus
 }
 #endif
```
```suggestion
#ifdef __cplusplus
}
namespace geopm
{
    /// @brief Verify a region's hint value is legal for use.
    void check_hint(uint64_t hint);
}
#endif
```

cmcantalupo replied:
The sole purpose of check_hint is to throw exceptions, so lets make sure it is only called in C++ contexts.

avilcheslopez replied:
Yep, it's only being used in C++ contexts.

cmcantalupo commented on src/ApplicationStatus.cpp:
```diff
@@ -61,10 +60,10 @@ namespace geopm
             throw Exception("ApplicationStatusImp::set_hint(): invalid CPU index: " + std::to_string(cpu_idx),
                             GEOPM_ERROR_INVALID, __FILE__, __LINE__);
         }
-        geopm::check_hint(hint);
```
lets leave the namespace in there...other similar changes below.

cmcantalupo commented on service/src/geopm_hint.h:
```diff
@@ -15,28 +15,31 @@ extern "C" {
+    GEOPM_REGION_HINT_IGNORE,   /* Do not add region time to epoch */
+    GEOPM_REGION_HINT_INACTIVE, /* Used to marke CPUs that are not running application */
+    GEOPM_SENTINEL_REGION_HINT,
+    GEOPM_MASK_REGION_HINT =    GEOPM_REGION_HINT_UNKNOWN |
+                                GEOPM_REGION_HINT_COMPUTE |
+                                GEOPM_REGION_HINT_MEMORY |
+                                GEOPM_REGION_HINT_NETWORK |
+                                GEOPM_REGION_HINT_IO |
+                                GEOPM_REGION_HINT_SERIAL |
+                                GEOPM_REGION_HINT_PARALLEL |
+                                GEOPM_REGION_HINT_IGNORE |
+                                GEOPM_REGION_HINT_INACTIVE,
```
This mask is no longer valid now that we aren't using bit hits as bitmasks.

avilcheslopez replied:
Yep, my bad, forgot to circle back to delete the mask enum.

cmcantalupo commented on src/Profile.hpp:
```diff
@@ -53,7 +53,7 @@ static inline uint64_t geopm_region_id_hint(uint64_t region_id)
     }
     else {
-        ret = region_id & GEOPM_MASK_REGION_HINT;
+        ret = (region_id >> 32) & GEOPM_MASK_REGION_HINT;
```
no need for masking anymore.

cmcantalupo reviewed (commented):
Mostly pretty good.  Just want to make sure check_hint is only called in C++, and we don't really have a use for the hint mask anymore.

avilcheslopez committed:
Misc. changes/fixes related to check_hint()

- Placed check_hint() in geopm namespace
- Use fully qualified name when invoking check_hint()
- Removed GEOPM_MASK_REGION_HINT as it's no longer needed

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Rename GEOPM_SENTINEL_REGION_HINT to GEOPM_NUM_REGION_HINT

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>
